### PR TITLE
Use collector to set migration metrics

### PR DESCRIPTION
**Pull request by machadovilaca**
**What this PR does / why we need it**:

Sometimes migration metrics `kubevirt_migrate_vmi_pending_count`, `kubevirt_migrate_vmi_scheduling_count` and `kubevirt_migrate_vmi_running_count` show negative values on migrating the VM.

I refactored the code to follow another approach using a VirtualMachineInstanceMigration collector.
I think this new alternative also goes in line with some concerns raised when these metrics were added https://github.com/kubevirt/kubevirt/pull/7465#issuecomment-1120891394, related to performing monitoring logic in the controller logic.

**Special notes for your reviewer**:

PR is marked as `size/XXL` because we removed some dependencies which caused a cleanup update in the vendor code

**Release note**:

```release-note
Use collector to set migration metrics
```

### CODE REVIEW

File: pkg/monitoring/migrationstats/collector.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+			"Number of failed migrations.",
+			[]string{"vmi", "vmim"},
+			nil,
+		),
+	}
+)
+
+type MigrationCollector struct {
+	migrationInformer cache.SharedIndexInformer
+	clusterConfig     *virtconfig.ClusterConfig
```
Comment (fossedihelm) [inline pkg/monitoring/migrationstats/collector.go:77]: I did not see any usage of the `clusterConfig`. Is it a leftover? Thanks

File: pkg/monitoring/migrationstats/collector.go [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+		clusterConfig:     clusterConfig,
+	}
+
+	prometheus.MustRegister(co)
+	return co
+}
+
+func (co *MigrationCollector) Collect(ch chan<- prometheus.Metric) {
+	cachedObjs := co.migrationInformer.GetIndexer().List()
+	vmims := make([]*k6tv1.VirtualMachineInstanceMigration, len(cachedObjs))
```
Comment (fossedihelm) [inline pkg/monitoring/migrationstats/collector.go:97]: How about adding a check that `return` if the length of cachedObjs is 0?

File: pkg/monitoring/migrationstats/collector_test.go [hunk 3]
```diff
@@ -20,11 +20,66 @@
...
+var _ = Describe("Collector", func() {
+	var ch chan prometheus.Metric
+	var scrapper *prometheusScraper
+
+	BeforeEach(func() {
+		ch = make(chan prometheus.Metric, 5)
+		scrapper = &prometheusScraper{ch: ch}
+	})
+
+	getVMIM := func(status k6tv1.VirtualMachineInstanceMigrationPhase) *k6tv1.VirtualMachineInstanceMigration {
```
Comment (fossedihelm) [inline pkg/monitoring/migrationstats/collector_test.go:43]: nit: rename status --> phase

File: pkg/monitoring/migrationstats/collector_test.go [hunk 4]
```diff
@@ -20,11 +20,66 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
+	io_prometheus_client "github.com/prometheus/client_model/go"
+
+	k6tv1 "kubevirt.io/api/core/v1"
 )
 
-var _ = BeforeSuite(func() {
-})
+var _ = Describe("Collector", func() {
```
Comment (fossedihelm) [inline pkg/monitoring/migrationstats/collector_test.go:34]: nit: I would prefer "Migration Stats Collector"

**Review by fossedihelm — COMMENTED** [review]
Thanks for your PR! Great to see this.
I left few comments below and some nits.
I would also suggest adding more unit tests in order to cover all the cases:
- PendingMigrations
- SchedulingMigrations
- RunningMigrations
- SucceededMigrations
- FailedMigrations

File: tests/monitoring/monitoring.go [hunk 5]
```diff
@@ -244,6 +258,37 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
...
+						{
+							MatchExpressions: []k8sv1.NodeSelectorRequirement{
+								{Key: "kubernetes.io/hostname", Operator: k8sv1.NodeSelectorOpIn, Values: []string{nodeName}},
+							},
+						},
+					},
+				},
+			},
+		}
+	}
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/monitoring.go]: changed

File: tests/monitoring/monitoring.go [hunk 6]
```diff
@@ -244,6 +258,37 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
...
+			if err != nil {
+				return err
+			}
+
+			if migration.Status.Phase != v1.MigrationFailed {
+				return fmt.Errorf("waiting on migration with phase %s to reach phase Failed", migration.Status.Phase)
+			}
+			return nil
+		}, 2*time.Minute, 5*time.Second).Should(Succeed(), "migration creation should fail")
+	}
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/monitoring.go]: awesome, changed

File: tests/monitoring/monitoring.go [hunk 7]
```diff
@@ -252,6 +297,76 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
...
+
+		AfterEach(func() {
+			revertScale(virtOperator.deploymentName)
+			time.Sleep(30 * time.Second)
+		})
+
+		It("Should correctly update metrics on successful VMIM", func() {
+			By("Creating VMIs")
+			vmi := tests.NewRandomFedoraVMIWithGuestAgent()
+			vmi = tests.RunVMIAndExpectLaunch(vmi, 240)
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/monitoring.go:286]: uniformized all with 240

File: tests/monitoring/monitoring.go [hunk 8]
```diff
@@ -481,15 +598,15 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
 				tests.ConfirmVMIPostMigration(virtClient, vmi, migrationUID)
 			}
 
+			By("Verifying KubeVirtVMIExcessiveMigration alert exists")
+			verifyAlertExist("KubeVirtVMIExcessiveMigrations")
+
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/monitoring.go:570]: Verify the metrics and alerts are updated before cleaning up all the resources

File: tests/monitoring/prometheus_utils.go [hunk 9]
```diff
@@ -90,6 +79,63 @@ func getMetricValue(cli kubecli.KubevirtClient, query string) (string, error) {
...
+	}
+
+	if s, ok := returnObj.(string); ok {
+		returnVal = s
+	} else {
+		return "", fmt.Errorf("metric value is not string")
+	}
+
+	return returnVal, nil
+}
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/prometheus_utils.go]: updated

File: pkg/monitoring/migrationstats/collector.go (hunk 1 above)
Comment (machadovilaca) [reply pkg/monitoring/migrationstats/collector.go]: removed

File: pkg/monitoring/migrationstats/collector.go (hunk 2 above)
Comment (machadovilaca) [reply pkg/monitoring/migrationstats/collector.go:96]: added

File: pkg/monitoring/migrationstats/collector_test.go (hunk 3 above)
Comment (machadovilaca) [reply pkg/monitoring/migrationstats/collector_test.go]: renamed

File: pkg/monitoring/migrationstats/collector_test.go (hunk 4 above)
Comment (machadovilaca) [reply pkg/monitoring/migrationstats/collector_test.go]: changed

Comment (machadovilaca) [reply pkg/monitoring/migrationstats/collector_test.go]: changed

**Comment by machadovilaca** [comment]
@fossedihelm I updated the PR with the requested changes and added unit tests for all cases

**Review by fossedihelm — COMMENTED** [review]
@machadovilaca Thanks! Looks good to me!

/lgtm

**Comment by enp0s3** [comment]
@machadovilaca Hi. Can you please elaborate on the following:
- How does the new refactoring better then the current approach
- Did the refactoring resolves the negative value issue?

File: hack/prom-rule-ci/prom-rules-tests.yaml [hunk 10]
```diff
@@ -701,20 +701,17 @@ tests:
...
-        values: "_ _ 1 2 2 _ _ 1 2 _ 1  2  2  2+0x13"        # 6 increases, in samples: 2,3,7,8,10,11
+      - series: 'kubevirt_migrate_vmi_succeeded_total{vmi="vmi-example-1", vmim="vmi-example-1-migration"}'
+        # time:  0 1 2 3 4 5
+        values: "_ _ _ _ _ 13"
 
     alert_rule_test:
-      # at 11h, there are total of 11 migrations made on a single VMI, so the alert should not be fired.
-      - eval_time: 11h
+      # at 4h, there are total of 11 migrations made on a single VMI, so the alert should not be fired.
+      - eval_time: 4h
```
Comment (enp0s3) [inline hack/prom-rule-ci/prom-rules-tests.yaml:710]: Why did the evaluation time changed?
Perhaps it deserve a separate commit with an explanation, because this is actually a functionality change rather than a refactoring

File: hack/prom-rule-ci/prom-rules-tests.yaml [hunk 11]
```diff
@@ -701,20 +701,17 @@ tests:
...
         alertname: KubeVirtVMIExcessiveMigrations
         exp_alerts: []
-      # at 16h, there are total of 13 migrations made on a single VMI, thus the alert is expected to be fired.
-      - eval_time: 16h
+      # at 5h, there are total of 13 migrations made on a single VMI, thus the alert is expected to be fired.
```
Comment (enp0s3) [inline hack/prom-rule-ci/prom-rules-tests.yaml:713]: same here

File: hack/prom-rule-ci/prom-rules-tests.yaml [hunk 12]
```diff
@@ -738,7 +735,7 @@ tests:
               kubernetes_operator_part_of: "kubevirt"
               kubernetes_operator_component: "kubevirt"
               vmi: vmi-example-1
-      # since the first increase occurred in 2h, will need to evaluate at 26h to disregard the first increase and clear the alert.
-      - eval_time: 26h
+      # since the first increase occurred in 5h, will need to evaluate at 30h to disregard the first increase and clear the alert.
```
Comment (enp0s3) [inline hack/prom-rule-ci/prom-rules-tests.yaml:738]: same here

File: hack/prom-rule-ci/prom-rules-tests.yaml [hunk 13]
```diff
@@ -701,20 +701,17 @@ tests:
   # Simulating metric resets at time slots 5 and 9
   - interval: 1h
     input_series:
-      - series: 'kubevirt_migrate_vmi_succeeded_total{vmi="vmi-example-1", source="node-1", target="node-2"}'
```
Comment (enp0s3) [inline hack/prom-rule-ci/prom-rules-tests.yaml:704]: Can you please explain why this test was changed? Also if the change is reasonable, the upper comment needs to be changed

File: tests/monitoring/monitoring.go [hunk 14]
```diff
@@ -281,6 +290,79 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
 		})
 	})
 
+	Context("VM migration metrics", func() {
+		var nodes *k8sv1.NodeList
+
+		BeforeEach(func() {
+			scales = make(map[string]*autoscalingv1.Scale, 1)
+			backupScale(virtOperator.deploymentName)
```
Comment (enp0s3) [inline tests/monitoring/monitoring.go:298]: Can you please elaborate on the reason you need to scale the virt-operator in this scenario?

File: hack/prom-rule-ci/prom-rules-tests.yaml (hunk 13 above)
Comment (machadovilaca) [reply hack/prom-rule-ci/prom-rules-tests.yaml:704]: The alert expression in `pkg/virt-operator/resource/generate/components/prometheus.go` was not correct, and the input series in this test is wrong.

`3+0x13` as used below does not expands into `2,4,6,7,11,12,13` but to `3` repeated 14 times, as explained in [Prometheus unit tests documentation](https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/#series).

Same for `2+0x13`.

Comment updated.

File: hack/prom-rule-ci/prom-rules-tests.yaml (hunk 10 above)
Comment (machadovilaca) [reply hack/prom-rule-ci/prom-rules-tests.yaml:710]: After the change explained above, I just moved the times back. I think it makes no difference, everything just happens sooner.

I'll move the alert changes into a separate commit.

File: tests/monitoring/monitoring.go (hunk 14 above)
Comment (machadovilaca) [reply tests/monitoring/monitoring.go]: Removed

Comment (machadovilaca) [reply tests/monitoring/monitoring.go]: Removed

**Comment by machadovilaca** [comment]
@enp0s3

> @machadovilaca Hi. Can you please elaborate on the following:
> 
> * How does the new refactoring better then the current approach

In the current approach I think we have 2 issues. First, the update of the metrics as is is not idempotent, and since the `updateStatus` function is called while processing elements of a queue, it might be executed more than 1 time, derived from migrations being requeued. Secondly, I think the current approach is much more complicated and harder to understand and maintain since we are updating metrics all around the migration controller's core logic.

With the proposed approach the collector reflects the state in the cluster and executing it an arbitrary number of times over a given state, will always produce the same output in Prometheus. Also, we separate the metrics logic from the migration controller's logic, which makes it simpler, easier to understand and maintain.

> * Did the refactoring resolves the negative value issue?

Yes. With the changes in this PR, metrics will be a straightforward representation of the state in the cluster.

File: tests/monitoring/monitoring.go [hunk 15]
```diff
@@ -323,23 +395,23 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
...
-			revertScale(virtApi.deploymentDame)
-			revertScale(virtController.deploymentDame)
-			revertScale(virtOperator.deploymentDame)
+			revertScale(virtApi.deploymentName)
+			revertScale(virtController.deploymentName)
+			revertScale(virtOperator.deploymentName)
 
 			time.Sleep(10 * time.Second)
 			waitUntilAlertDoesNotExist("VirtOperatorDown")
 			waitUntilAlertDoesNotExist("NoReadyVirtOperator")
```
Comment (enp0s3) [inline tests/monitoring/monitoring.go:410]: Since I follow your improvement logic, why don't you add
```
			waitUntilAlertDoesNotExist(virtController.downAlert)
			waitUntilAlertDoesNotExist(virtApi.downAlert)
```
here as well?

File: tests/monitoring/monitoring.go [hunk 16]
```diff
@@ -392,11 +464,13 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
 			if !errors.IsAlreadyExists(err) {
 				util.PanicOnError(err)
 			}
-			revertScale(virtOperator.deploymentDame)
+			revertScale(virtOperator.deploymentName)
 
 			time.Sleep(10 * time.Second)
-			waitUntilAlertDoesNotExist("VirtOperatorDown")
-			waitUntilAlertDoesNotExist("NoReadyVirtOperator")
+			waitUntilAlertDoesNotExist(virtOperator.downAlert)
```
Comment (enp0s3) [inline tests/monitoring/monitoring.go:470]: I see that you'e omitted `waitUntilAlertDoesNotExist("NoReadyVirtOperator")`, is that in purpose?

File: tests/monitoring/monitoring.go [hunk 17]
```diff
@@ -281,6 +290,69 @@ var _ = Describe("[Serial][sig-monitoring]Prometheus Alerts", func() {
 		})
 	})
 
+	Context("VM migration metrics", func() {
+		var nodes *k8sv1.NodeList
+
+		BeforeEach(func() {
```
Comment (enp0s3) [inline tests/monitoring/monitoring.go:296]: I suggest to add `checks.SkipIfMigrationIsNotPossible()` since its being widely used in the e2e suites where migration should be tested.

File: hack/prom-rule-ci/prom-rules-tests.yaml (hunk 13 above)
Comment (enp0s3) [reply hack/prom-rule-ci/prom-rules-tests.yaml:704]: TBH I don't understand what's wrong with the expression, please see https://github.com/kubevirt/kubevirt/pull/7757 discussion, me and Barak verified it. Can please elaborate on the error?

Comment (enp0s3) [reply hack/prom-rule-ci/prom-rules-tests.yaml:704]: Does this change/fix related to the scope of the PR?

File: tests/monitoring/prometheus_utils.go [hunk 18]
```diff
@@ -90,6 +82,41 @@ func getMetricValue(cli kubecli.KubevirtClient, query string) (string, error) {
...
+	return true
+}
+
+func fetchMetric(cli kubecli.KubevirtClient, query string) (QueryRequestResult, error) {
+	bodyBytes := DoPrometheusHTTPRequest(cli, fmt.Sprintf("/query?query=%s", query))
+
+	var result QueryRequestResult
+	err := json.Unmarshal(bodyBytes, &result)
+	if err != nil {
+		return QueryRequestResult{}, err
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/prometheus_utils.go]: changed, changed the return type to a pointer

File: tests/monitoring/prometheus_utils.go [hunk 19]
```diff
@@ -90,6 +82,41 @@ func getMetricValue(cli kubecli.KubevirtClient, query string) (string, error) {
...
+	}
+
+	if result.Status != "success" {
+		return QueryRequestResult{}, fmt.Errorf("api request failed. result: %v", result)
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/prometheus_utils.go]: changed

File: tests/monitoring/prometheus_utils.go [hunk 20]
```diff
@@ -90,6 +82,41 @@ func getMetricValue(cli kubecli.KubevirtClient, query string) (string, error) {
...
+
+	return nil
+}
+
+func labelsMatch(pr promResult, labels map[string]string) bool {
+	for k, v := range labels {
+		if pr.Metric[k] != v {
+			return false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/monitoring/prometheus_utils.go:100]: updated?
is there any way we can add this rule to the linter?

File: tests/monitoring/monitoring.go (hunk 15 above)
Comment (machadovilaca) [reply tests/monitoring/monitoring.go]: added

File: tests/monitoring/monitoring.go (hunk 16 above)
Comment (machadovilaca) [reply tests/monitoring/monitoring.go:474]: in this set of tests, I think it's okay to skip that verification

File: tests/monitoring/monitoring.go (hunk 17 above)
Comment (machadovilaca) [reply tests/monitoring/monitoring.go:296]: added

File: hack/prom-rule-ci/prom-rules-tests.yaml (hunk 13 above)
Comment (machadovilaca) [reply hack/prom-rule-ci/prom-rules-tests.yaml:704]: `increase(v)` is syntactic sugar for `rate(v)` multiplied by the number of seconds. This is problematic because it uses extrapolation in its computation and it produces variable results depending on the time intervals. It's possible to easily break the expression in Prometheus unit tests with:

```yaml
  - interval: 10s
    input_series:
      - series: 'kubevirt_migrate_vmi_succeeded_total{vmi="vmi-example-1", source="node-2", target="node-1"}'
        # time:  0 10 20 30 40 50 60
        values: "0 12 12 12 12 12 12"

    alert_rule_test:
      - eval_time: 30s
        alertname: KubeVirtVMIExcessiveMigrations
        exp_alerts:
          - exp_annotations:
              description: "VirtualMachineInstance vmi-example-1 has been migrated more than 12 times during the last 24 hours"
              summary: "An excessive amount of migrations have been detected on a VirtualMachineInstance in the last 24 hours."
              runbook_url: "https://kubevirt.io/monitoring/runbooks/KubeVirtVMIExcessiveMigrations"
            exp_labels:
              severity: "warning"
              kubernetes_operator_part_of: "kubevirt"
              kubernetes_operator_component: "kubevirt"
              vmi: vmi-example-1
```

Result:

```
  FAILED:
    alertname:KubeVirtTest, time:30s, 
        exp:"[Labels:{alertname=\"KubeVirtVMIExcessiveMigrations\", kubernetes_operator_component=\"kubevirt\", kubernetes_operator_part_of=\"kubevirt\", severity=\"warning\", vmi=\"vmi-example-1\"} Annotations:{description=\"VirtualMachineInstance vmi-example-1 has been migrated more than 12 times during the last 24 hours\", runbook_url=\"https://kubevirt.io/monitoring/runbooks/KubeVirtVMIExcessiveMigrations\", summary=\"An excessive amount of migrations have been detected on a VirtualMachineInstance in the last 24 hours.\"}]", 
        got:"[]"

```

That being said, the problem only happens within specific timings and if we reevaluate again for example at 60s, behavior is the expected.

File: pkg/virt-operator/resource/generate/components/prometheus.go [hunk 21]
```diff
@@ -486,8 +486,7 @@ func NewPrometheusRuleSpec(ns string, workloadUpdatesEnabled bool) *v1.Prometheu
 					},
 					{
 						Alert: "KubeVirtVMIExcessiveMigrations",
-						Expr:  intstr.FromString("floor(increase(sum by (vmi) (kubevirt_migrate_vmi_succeeded_total)[1d:1m])) >= 12"),
-						For:   "1m",
+						Expr:  intstr.FromString("sum by (vmi) (max_over_time(kubevirt_migrate_vmi_succeeded_total[1d])) >= 12"),
```
Comment (enp0s3) [inline pkg/virt-operator/resource/generate/components/prometheus.go:489]: @machadovilaca As discussed, please check whether using the extra labels is needed for the succeeded migrations metric.

File: tests/libvmi/vmi.go [hunk 22]
```diff
@@ -184,6 +184,24 @@ func WithCPUFeature(featureName, policy string) Option {
 	}
 }
 
+func WithNodeAffinityFor(node *k8sv1.Node) Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		vmi.Spec.Affinity = &k8sv1.Affinity{
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tests/libvmi/vmi.go]: updated

File: pkg/monitoring/migrationstats/collector.go [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+
+	prometheus.MustRegister(co)
+	return co
+}
+
+func (co *MigrationCollector) Collect(ch chan<- prometheus.Metric) {
+	cachedObjs := co.migrationInformer.GetIndexer().List()
+	if len(cachedObjs) == 0 {
+		return
+	}
```
Comment (enp0s3) [inline pkg/monitoring/migrationstats/collector.go:95]: missing new line after `if`

File: pkg/monitoring/migrationstats/collector.go [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+		desc,
+		prometheus.GaugeValue,
+		value,
+		labelValues...,
+	)
+	if err != nil {
+		log.Log.V(4).Warningf("Error creating the new const metric for %s: %s", desc, err)
+		return
+	}
+	ps.ch <- mv
```
Comment (enp0s3) [inline pkg/monitoring/migrationstats/collector.go:149]: missing new line after `if`

File: pkg/monitoring/migration/prometheus.go [hunk 25]
```diff
@@ -21,133 +21,13 @@ package migration
...
-	incMigrationCounter(vmi, targetPod, MigrationsFailedTotal)
-}
-
-func getMigrationSourceAndTarget(vmi *virtv1.VirtualMachineInstance, targetPod *k8sv1.Pod) (source, target string) {
-	source = vmi.Status.NodeName
-	if targetPod != nil {
-		target = targetPod.Spec.NodeName
-	}
-
-	if vmi.Status.MigrationState != nil {
```
Comment (enp0s3) [inline pkg/monitoring/migration/prometheus.go:129]: @machadovilaca You could preserve the source and target nodes information in the new logic, by adding a VMI informer, then reading the the VMI status using the VMI name from the VMIM. Is there a reason you decided to drop the source and target node labels?

File: tests/monitoring/prometheus_utils.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (enp0s3) [reply tests/monitoring/prometheus_utils.go:100]: @machadovilaca Yes I am willing to add that. Waiting for some free BW

**Review by enp0s3 — COMMENTED** [review]
Looks good, last round and I think we are good

File: tools/doc-generator/fakeMigrationsCollector.go [hunk 26]
```diff
@@ -0,0 +1,39 @@
+package main
```
[the remark this answers is not in the document]
Comment (machadovilaca) [reply tools/doc-generator/fakeMigrationsCollector.go:1]: added

File: pkg/monitoring/migrationstats/collector.go (hunk 23 above)
Comment (machadovilaca) [reply pkg/monitoring/migrationstats/collector.go:95]: added

File: pkg/monitoring/migrationstats/collector.go (hunk 24 above)
Comment (machadovilaca) [reply pkg/monitoring/migrationstats/collector.go:149]: added

File: pkg/monitoring/migration/prometheus.go (hunk 25 above)
Comment (machadovilaca) [reply pkg/monitoring/migration/prometheus.go:129]: We are not using that information in any alerts or recording rules, and since in this PR I added the VMIM name information in the labels, it's very easy for the user to get the source and target node information starting from the VMIM name

**Review by fossedihelm — APPROVED** [review]
Thanks

**Comment by kbidarkar** [comment]
@machadovilaca Feel we need a backport to release 0.58.0

**Comment by fossedihelm** [comment]
@machadovilaca automatic cherry-pick fails. So you need to do it manually. Sorry for this :(

**Comment by machadovilaca** [comment]
@fossedihelm created the backport PR
